engine/consensus/compliance/engine.go

```diff
@@ -236,6 +195,31 @@ func (e *Engine) OnFinalizedBlock(block *model.Block) {
 	}
 }
 
+// OnBlockProposal feeds a new block proposal into the processing pipeline.
+// Incoming proposals are queued and eventually dispatched by worker.
+func (e *Engine) OnBlockProposal(proposal *messages.BlockProposal) {
+	e.core.engineMetrics.MessageReceived(metrics.EngineCompliance, metrics.MessageBlockProposal)
+	if e.pendingBlocks.Push(inboundBlock{proposal.Header.ProposerID, proposal}) {
```

Comment:
Fundamentally, I agree with Jordan.

These thoughts might be tangental, but maybe helpful for clarifying things
* there is an important difference between blocks that we receive through gossip vs blocks we receive through the synchronization engine: 
   * for gossiped blocks, the networking layer can give me a proof that the proposer originally sent this block
   * for blocks I receive from the synchronization engine, the network layer can only tell me who relayed the block
  
   This is ok, because in either case, I can hold the origin accountable for message:
    * for gossiped proposals, we hold the proposer accountable
    * for synchronized blocks, the protocol mandates that the party who feeds the block to me is only sending valid blocks

The information about the message origin containing the block would be carried by the inspector object, which would be replacing the insufficient notion of `originID`. But until we implement message forensics, `originID` is our placeholder for the inspector object. 

Therefore, I would advocate the use `originID` now everywhere, were we would be using the inspector object later. This means that the MessageHub should pass the `originID` along the message to the compliance engine.

Reply:
@AlexHentschel @jordanschalm Let me know what you think about this approach: https://github.com/onflow/flow-go/pull/3512/commits/4c8276a441087d267746ab44af1f58ebd3a1313c. 
I've got rid from `inboundBlock` and started using `events.SyncedClusterBlock` for all events. Works pretty nice in my opinion. If you are good with it I will make the same change for main consensus.

Reply:
That approach looks good to me, I would only suggest to update the documentation for `SyncedClusterBlock` (maybe name too) to be clear that it is an internal representation for a block + block sender, regardless of how we got the block. Currently it represents specifically blocks obtained through the sync protocol.

Reply:
I'd like to take a step back and ask the question: Why do we need a dedicated `events.SyncedClusterBlock` struct? 

My thoughts on the matter
* we desire a single struct so we can queue the inbound message: https://github.com/onflow/flow-go/blob/4c8276a441087d267746ab44af1f58ebd3a1313c/engine/common/follower/engine.go#L163-L170
  We have similar requirements for other engines that process untrusted inbound messages.
  But needing a compound struct for the purpose of queuing is an internal implementation aspect of the various engines. The APIs themselves are strongly typed, special purpose functions: https://github.com/onflow/flow-go/blob/4c8276a441087d267746ab44af1f58ebd3a1313c/engine/collection/compliance/engine.go#L172 
* the block being unverified is the main reason, why we need to retain information about the origin. Once the block is verified, we know that that it initially originated from the proposer and don't care whether we got it through gossip from the proposer or via the sync engine. I feel this is a pretty general pattern, which applies to many messages, not just blocks. 
* Currently, the struct definitions are in [`model/events/synchronization.go`](https://github.com/onflow/flow-go/blob/4c8276a441087d267746ab44af1f58ebd3a1313c/model/events/synchronization.go#L8-L16). Note that they are the _only_ structs in this package. The naming suggests some relation to synchronization, which I feel is not the case. 

This leads me to the following suggestion:
* Instead of implementing the same code for many messages individually, we could utilize Generics to cover this for all messages. My proposal would be to include the following struct in the package `model.flow` 
    ```golang
    // Slashable represents an message we got from a different node, whose validity has not been _entirely_
    // confirmed. We want to retain information about the origin that published this particular message
    // within the network, so we can potentially raise a slashing challenge against the origin, should
    // we discover that this message is evidence of a protocol violation.
    // TODO: this struct does not have its final form. We only retain the ID of the node where the message
    // originated. However, this will not allow us to prove to somebody else that `OriginID` really sent
    // this message, as we could have tempered with the message and then erroneously accuse `OriginID`
    // of a protocol violation. In the mature protocol, `OriginID` will be replaced by an inspector object
    // (generated by the networking layer) that allows us to generate a cryptographic proof of who sent the message.
    type Slashable[T any] struct {
    	OriginID Identifier // this will become the inspector object, once we have message forensics
    	Message  *T
    }
    ```
*  the method https://github.com/onflow/flow-go/blob/4c8276a441087d267746ab44af1f58ebd3a1313c/engine/collection/compliance/engine.go#L198 could then change to 
    ```golang
    func (e *Engine) OnClusterBlockProposal(proposal flow.Slashable[messages.ClusterBlockProposal]) {
    ```

Reply:
Yeah, this a good idea, we could get rid of all repetitions of similar structures. I will implement it.